js/modules/k6/ws/ws.go

```diff
@@ -399,6 +242,72 @@ func (mi *WS) Connect(url string, args ...goja.Value) (*HTTPResponse, error) {
 	}
 }
 
+func (mi *WS) dial(
+	ctx context.Context, state *lib.State, rt *goja.Runtime, url string,
+	args *wsConnectArgs,
+) (*Socket, *http.Response, func(), error) {
+	// Overriding the NextProtos to avoid talking http2
+	var tlsConfig *tls.Config
+	if state.TLSConfig != nil {
+		tlsConfig = state.TLSConfig.Clone()
+		tlsConfig.NextProtos = []string{"http/1.1"}
+	}
+
+	wsd := websocket.Dialer{
+		HandshakeTimeout: time.Second * 60, // TODO configurable
+		// Pass a custom net.DialContext function to websocket.Dialer that will substitute
+		// the underlying net.Conn with our own tracked netext.Conn
+		NetDialContext:    state.Dialer.DialContext,
+		Proxy:             http.ProxyFromEnvironment,
+		TLSClientConfig:   tlsConfig,
+		EnableCompression: args.enableCompression,
+		Jar:               args.cookieJar,
+	}
+	// this is needed because of how interfaces work and that wsd.Jar is http.Cookiejar
+	if args.cookieJar == nil {
+		wsd.Jar = nil
+	}
+
+	connStart := time.Now()
+	conn, httpResponse, dialErr := wsd.DialContext(ctx, url, args.headers)
+	connEnd := time.Now()
+
+	if state.Options.SystemTags.Has(metrics.TagIP) && conn.RemoteAddr() != nil {
+		if ip, _, err := net.SplitHostPort(conn.RemoteAddr().String()); err == nil {
+			args.tagsAndMeta.SetSystemTagOrMeta(metrics.TagIP, ip)
+		}
+	}
```

Comment:
I see we have this same code across http/grpc/ws should we create a helper function and reuse it?

Reply:
:thinking: Probably. But that would expand the scope of this refactor, and I didn't want to spend too much time on it.

Maybe create an issue for it?